internal/testutil/testutil_test.go

```diff
@@ -0,0 +1,71 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package testutil
+
+import (
+	"net"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestGetAvailableLocalAddress(t *testing.T) {
+	endpoint := GetAvailableLocalAddress(t)
```

Comment:
This looks like a race condition to me.
`GetAvailableLocalAddress()` calls `findAvailable()`, which calls `findAvailableAddress()`. `findAvailableAddress()` opens a generic TCP listener and closes it. The host and port part of this then closed listener is returned and in the next line of `TestGetAvailableLocalAddress()` a new listener is opened on this host/port combination.

A possible race condition can happen between the closing of the listener in `findAvailableAddress()` and before `GetAvailableLocalAddress()` returns and a new listener is opened on the returned endpoint.

Reply:
Nice catch, I just noticed this comment in the `findAvailableAddress` :

```
// There is a possible race if something else takes this same port before
// the test uses it, however, that is unlikely in practice.
```

Although we could add some sync protection, the `GetAvailableLocalAddress` function will be used across multiple packages and as it does not return a listener, there can be a race condition at the moment of its usage.

The upstream collector has been using this helper for a while (and many more usages) without a noticeable collision. Adding a reference to the upstream code in case it evolves.

Reply:
`GetAvailableLocalAddress` is only intended for tests, right? So the intended usage is just to get an available address for tests, if the race condition happens, the only consequence is that the test will fail, because the address may be taken when in the test we'd really use it. Or I'm I mistaken?